### ISSUE TITLE
Support for Futures markets

### ISSUE THREAD

**Issue by andre77** [knowm/XChange#3171]
Currently the central key point of the XChange API is the CurrencyPair, which is passed as parameter for almost all methods. This works fine so far for spot markets. But if we want to support futures markets the CurrencyPair is not enough, for example see [Deribit Instruments](https://www.deribit.com/api/v2/public/get_instruments?currency=BTC&expired=false&kind=future), there we have three "products" with the same pair BTC/USD.
To support this we would need to replace the CurrencyPair by `Product` class, which could look like this:

```
public class Product {
  private String base;
  private String counter;
  private String symbol;	// this can be null, ie for spot products
  private ProductType productType;
}

public enum ProductType {
  spot, perpetual, future;
}
```

To stay with the deribit example, we would have three instances for products:
```
{base: BTC, counter: USD, symbol: BTC-27SEP19, productType: future}
{base: BTC, counter: USD, symbol: BTC-27DEC19, productType: future}
{base: BTC, counter: USD, symbol: BTC-PERPETUAL, productType: perpetual}
```

So like already said, inside the interface methods we could replace the CurrencyPair by Product (or Instrument or what ever name). The old methods could actually stay for a while and be marked as deprecated. This way we would be still compatible with older versions.

if the majority thinks it could make sense, i will create a PR

**Comment by makarid** [comment]
I like the idea. Maybe we can add options support also. They just need expiration date field i believe. What do you thing?

**Comment by andre77** [comment]
futures also have expiration date, i do not mind to add an optional Date field, which would be null for 'spot' and 'perpetual' products

**Comment by makarid** [comment]
Great.

**Comment by timmolter** [comment]
sounds good to me.

**Comment by makarid** [comment]
Have anyone implement this or at least started?

**Comment by andre77** [comment]
not yet unfortunately

**Comment by douggie** [comment]
@andre77 wondered if you had any thoughts on the product PR idea, just I was going to look at implementing TFX which would need a simliar solution, I was going to approach it a bit different but may be your way is better. 

https://github.com/knowm/XChange/issues/3258

**Comment by andre77** [comment]
no, unfortunately did not find any time to make some progress here

**Comment by mdvx** [comment]
>> Maybe we can add options support also. They just need expiration date field and strike price i believe.We can create a second constructor for option support. 

Vanilla options can be covered this way, but any Complex option will need more (for instance, is it American, European or Asian style (multiple expiry dates)), and there are Binary options etc...  It may be better to add a hashMap to the Product to support all the permutations than can be populated from ExchangeMetaData  (local or downloaded).  

I am not sure we want to build a normalized product database here, but allowing products to have un-restricted properties would be a great halfway house.

**Comment by mdvx** [comment]
see Month Codes, for instance Z for December might be useful for sorting.  These codes refer to the 3rd Friday in the month, but I see from the data above BTC-27DEC19 and that is the 4th Friday, expiry date needs to be kept.

https://www.cmegroup.com/education/courses/introduction-to-futures/understanding-contract-trading-codes.html

**Comment by douggie** [comment]
how about we introduce an abstract instrument class, which the existing Currency pair extends, then each exchange implentation can extend the instrument class as they needed for any particular exchange implentation. just think this would reduce the amount of refactoring.

**Comment by douggie** [comment]
@andre77 @makarid   looking at this a bit more may be abstract classes say 'Instrument' with two concrete classes of 'CurrencyPair' and 'Contract' might create more Hassel in the JSON serializers that using concrete classes. To avoid a large refactoring what are the thoughts of creating a 'Contract' class that extends 'CurrencyPair' which assumes any derviative/contract would be FX style with a base and quote currency?

**Comment by mdvx** [comment]
I would avoid extending CurrencyPair, and wrap it instead.  Because some contract will be defined by multiple CurrencyPair objects.

**Comment by douggie** [comment]
Sure, so may be I will take @andre77 original suggestion and itroduce say an 'instrument' class that wraps currency pair and update all the interfaces to use new instrument class.

**Comment by makarid** [comment]
> Sure, so may be I will take @andre77 original suggestion and introduce a new concrete class say called 'instrument' that wraps currency pair and update all the interfaces to use new instrument class.

We could first create a new interface Future.class and sub divide it to FutureAccountService,FutureTradingService,FutureMarketDataService and implement these to the exchanges that support futures. That way we could create the Instrument.class and only add it the methods of these interfaces. That way we don't need to modify the existing classes.

**Comment by walec51** [comment]
> 'Instrument' with two concrete classes of 'CurrencyPair' and 'Contract' might create more hassle in the JSON serialisers that using concrete classes

Jackson can handle far more complex class hierarchies then this - you just have to know how to configure it via annotations - this is not a problem

> FutureAccountService,FutureTradingService,FutureMarketDataService

would those interfaces be any different from the normal interfaces other then from the fact that they accept Future instead of CurrencyPair?

what future specific methods would there be in a FutureAccountService?

what future specific methods would there be in a FutureMarketDataService?

do exchanges expose totally different APIs for futures?

**Comment by mdvx** [comment]
>> do exchanges expose totally different APIs for futures?
Yes, frequently (and they often replicate the spot API which changes for Futures)
for instance https://support.kraken.com/hc/en-us/categories/360001806372-Futures-API

Also, they can use different endpoints

**Comment by makarid** [comment]
> > 
> 
> what future specific methods would there be in a FutureAccountService?
> 
> what future specific methods would there be in a FutureMarketDataService?
> 

I don't know the exact methods yet. We could add them when we will try to implement at least one Future exchange in order to see what we will need. I was just posting about an idea of how we could solve the issue of adding Future support but without break things from the spot market exchanges. Definitely we would need at least the following methods which there aren't in existing  Xchange implementation(we can change the names of course):

getOpenPositions
getCurrentMarginUsed

**Comment by mdvx** [comment]
Where would leverage go? on the trade and position rather than the Instrument?  Maybe on the instrument to if leverage is fixed for that symbol.

Also, how to handle generic futures (March,June,September,December) vs dated contracts expiry (2020-03-27)?

**Comment by makarid** [comment]
> Where would leverage go? on the trade and position rather than the Instrument? Maybe on the instrument to if leverage is fixed for that symbol.
> 
> Also, how to handle generic futures (March,June,September,December) vs dated contracts expiry (2020-03-27)? Send in a trade for March-2020 get back executions for 2020-03-27?

There are 2 types of leverage. The instrument leverage(maximum leverage supported) and the openPosition leverage that is a real-time metric in order to see how much leverage your position is using.

I didn't understand the second question

**Comment by makarid** [comment]
The question that i would like to ask. Do we need to create a Position.class that extends Order.class or it is better to not extend Order.class? I am asking this because a Position is very different than an Order. A position metrics(margin,value) changing overtime but this is not true with an Order.

**Comment by makarid** [comment]
So, as i see it right now:

FutureAccountService.interface : contains { getAccountUsedLeverage() }
FutureTradingService.interface : contains { getOpenPositions(), getOpenPositions(CurrencyPair currencyPair, getFundingHistory(CurrencyPair currencyPair}
Also we need to add an OpenPosition.class. 

What do you think? I will start the implementation when we all agreed on something.

**Comment by mdvx** [comment]
I would not recommend extending the Order Class, I think it has already been overused (Trades and OrderBook) and the LimitOrder MarketOrder, StopOrder structure is not ideal.

**Comment by mdvx** [comment]
Generic futures codes are aliases to the nearest expiry month

**Comment by timmolter** [comment]
There are so many ways this could be done, which makes it hard to decide. We could add new DTOs, Services and methods. We could add new DTOs and methods to existing services. We could add new and/or extended DTOs to existing Services and existing methods. 

I always tried to lean towards making it easy for the end user (a programmer using the library, not the developers of the library) to understand the API so if we continue with this goal in mind, adding new DTOs and methods would make the most sense. Otherwise newbies will have to deal with complex "wrapper" classes for method arguments and return values. If someone comes and just want to place a limit order for a plain old "spot", they can easily find it by searching the methods and it is relatively easy then to understand what arguments are required and what is returned. As a compromise we could forgo adding new service classes but add more methods within the existing ones. By default the new methods will just throw a `NotImplementedYetException` as it always has worked.

I think this would warrant a version bump to 4.5.0. 

As far as extending `Order` or `CurrencyPair` or not I'd say yes if there is significant overlap of class members, but I'd say no if the members were completely or mostly orthogonal. Just take it by a case to case basis.

@mdvx If you think extending the Order class is not ideal, in which way do you think things could be improved?

**Comment by makarid** [comment]
Thank you @timmolter for your input. So, i will just add new DTOs and new methods to existing services.

**Comment by mdvx** [comment]
Here are some different concepts for Order class...

* Order -> OrderRequest
The Order class does not allow for enough parameters, flags was added to extend the functionality, but things like GTD need fields.  A MyExchangeOrder class can be created, but a lot of these are common across exchanges, and it would need to be applied to LimitOrder and StopOrder too.  An interface for GTD might be a solution. Another possibility is to add a Map to the base Order object for extra the fields, then GTD could go in the map.
GTD is just one such field, others might include LocateReqd flag, SettlCurrency, QuoteID, SettlType 

* Order -> OrderState of private Trade
We are missing an error status here for rejected Orders, whereas this is not required when placing synchronous REST calls, asynchronous calls need a place to report this.  REST are generally synchronous, but there is no reason an exchange would not reject an order further down its processing pipeline, maybe even after it is partially executed (risk limit hit after price moves)

* Order -> public Trade
This assumes a fully executed order, sometimes with maker and taker ids.

* LimitOrder -> OrderBookEntry
Problem here is the orderbook can combine many Orders for a single Entry (n<-1)

* Order -> Position
Well I thinks that is a lot to ask from one class

My tuppence.

**Comment by douggie** [comment]
With regards to extending the currency pair, I thought it might make sense
as some derivatives will have a base/quote and a prompt month such as
BTC.USD.PERPETUAL, even FTX's BTC-MOVE-1116 is really BTC.USD
volatility plus a prompt, (i.e BTC.USD.MOVE.1116) so it has an additional
symbol MOVE, where other derivatives such as TRUMP-2020
<https://ftx.com/trade/TRUMP>  or SHIT-PERP don't have an
underlying currency pair. So may be it makes more sense to have a new type
instryment that has an underlying that can be anything, most cases it will
be an CurrencyPair but could be something else like TRUMP or SHIT (shitcoin
index)

**Comment by mdvx** [comment]
Here are the 4 near (+1 prior) month CME contracts
BTC.G20 | Feb 2020 | 8825 | 8905 | 8430 | 8890 | -85 | -0.98% | 10:33
BTC.H20 | Mar 2020 | 8830 | 8910 | 8700 | 8775 | -25 | -0.28% | 17:28
BTC.J20 | Apr 2020 | 8940 | 8970 | 8795 | 8840 | -30 | -0.33% | 17:28
BTC.K20 | May 2020 | 9030 | 9070 | 8970 | 8970 | -20 | -0.22% | 17:28
BTC.M20 | Jun 2020 | 9150 | 9215 | 9035 | 9075 | -55 | -0.60% | 17:28

But we also have contracts on the spreads between two futures  (J,K, are six week codes)
BTC.G20:H20 | Feb 2020/Mar 2020 Spread | 41 | 80 | 41 | 77 | +35 | +51.47% | 10:33
BTC.H20:J20 | Mar 2020/Apr 2020 Spread | 85 | 90 | 80 | 80 | -12 | -17.65% | 17:28
BTC.H20:K20 | Mar 2020/May 2020 Spread | 201 | 201 | 190 | 190 | -12 | -5.97% | 17:28
BTC.H20:M20 | Mar 2020/Jun 2020 Spread | 314 | 314 | 300 | 300 | -25 | -7.96% | 17:28
BTC.J20:K20 | Apr 2020/May 2020 Spread | 103 | 111 | 101 | 108 | +2 | +1.82% | 17:28
BTC.J20:M20 | Apr 2020/Jun 2020 Spread | 199 | 252 | 199 | 225 | +55 | +27.64% | 17:28
BTC.K20:M20 | May 2020/Jun 2020 Spread | 103 | 105 | 103 | 105 | -14 | -13.59% | 17:28

https://quotes.ino.com/exchanges/contracts.html?r=CME_BTC

**Comment by mdvx** [comment]
here is one with 3 currencies 
Litecoin (BITCOIN:LTCUSD)

https://quotes.ino.com/exchanges/category.html?c=cryptocurrencies

**Comment by mdvx** [comment]
If we think to the defining of a Future contract, it is an obligation to Buy/Sell and asset at a future date.  Sometimes these contracts are settled in the quoted fiat, however it is possible for these contracts to be settled by 'physical' delivery, which would mean in actual BTC.  In which the USD part is just a reference and this very same contract could be quoted in JPY, EUR, or even ETH.

**Comment by douggie** [comment]
so based on above, how do people feel about Introduce a new type of instruments, say two new concert classes FuturesContract and OptionsContract (possibly extends futures contract as it will have a lot of the same attributes, plus type, strike etc). Futures contract which will have many attributes one of which is underlying, which could be a CurrencyPair or any other instrument we support, for example an option could have a underlier of a FuturesContact.

**Comment by mdvx** [comment]
OptionsContract should not extend FuturesContract as options can be on the underlying asset as well as the Future.  Simple Vanilla options will have a StrikePrice and an ExpiryDate, however they get a lot more complex, some with multiple ExpiryDates (Asian options), and many other types of derivations.

**Comment by walec51** [comment]
@mdvx according to what you said 

* FuturesContract should point to a CurrencyPair
* OptionsContract should point to an Instrument 

but this would allow an OptionsContract on an OptionsContract - could this be a thing in the real world? if not we can always ad some validation in the DTO

**Comment by douggie** [comment]
coolio, so just this then, where the Futures and Options contract can have an underlier of type Instrument. 
```
Instrument (Abstract Class)
 |___ CurrencyPair
 |___ FuturesContract
 |__OptionsContract

**Comment by mdvx** [comment]
> coolio, so just this then, where the Futures and Options contract can have an underlier of type Instrument.
> 
> ```
> Instrument (Abstract Class)
>  |___ CurrencyPair
>  |___ FuturesContract
>  |___ OptionsContract
> ```
This looks like the right structure, to me

**Comment by mdvx** [comment]
There was a question about Options on Options I saw last night.  These things do exists, as compound options, however they are the exceptional, and as such I would not be inclined to model them here.

https://www.investopedia.com/terms/c/compoundoption.asp

**Comment by douggie** [comment]
Don't think it matters, as both option contracts and futures contracts will
have an underlier of instrument which would allow for options on
currencypairs,  futures as well as on options.

**Comment by douggie** [comment]
@mdvx @walec51 anyone started looking at this on a branch? Happy to collaborate or I can start a branch early next week.

**Comment by walec51** [comment]
this https://github.com/knowm/XChange/pull/3461 would break the entire API so I doubt it will get merged

few exchanges have futures / options and therefore we talked to make either separate methods or even separate interfaces for them

wont the orders for options / futures have different fields and behaviors then the current `LimitOrder`, `MarketOrder` and `SpotOrder`?

if so then there is no sense in pretending that they can accept any thing other then a `CurrencyPair`

I would suggest to change the `Instument` class hierarchy to

```
Instrument (Abstract Class)
 |___ CurrencyPair
 |___ Derivative
     |___ FuturesContract
     |___ OptionsContract
```

and instead of breaking the current API to add methods (or even separate interfaces) that can handle `Derivative`s

**Comment by walec51** [comment]
@makarid in this PR https://github.com/knowm/XChange/pull/3439 suggested that the fields for and order on a derivative must by totally different

if he is wrong and orders for derivatives can be handled by our existing order type from the core API then I'm in favor of breaking the core API to introduce the Instrument class - however this will mean an entirely new version of our library and force all our users to refactor their code

**Comment by walec51** [comment]
so the crucial question we have to answer before we choose a direction is:

can our current core API `Order` (`LimitOrder`, `SpotOrder`, `MarketOrder`), `Trade` and `Ticker` classes be made to work for derivatives if we only introduce the `Instument` ? 

aren't there any specific fields that a derivative would have to have in an `Order`, `Trade` or `Ticker` class?

**Comment by douggie** [comment]
I am pretty sure the api's are very simliar, just the symbol changes to incorporate the pair and expiry.

**DEPTH**
**Karken**
`https://futures.kraken.com/derivatives/api/v3/orderbook?symbol=fi_xbtusd_180615`
`https://api.kraken.com/0/public/Depth?pair=XBTCHF`
**OKEX**
`https://www.okex.com/api/spot/v3/instruments/BTC-USDT/book?size=5&depth=0.2`
`https://www.okex.com/api/futures/v3/instruments/BTC-USD-180309/book?size=50`
**Binance**
`https://api.binance.com/api/v3/depth?symbol=ETHBTC`
`https://fapi.binance.com/fapi/v1/depth?symbol=ETHUSDT`


**Orders**
**Binanace**
`https://api.binance.com/api/v3/order{"symbol":"ETHBTC","side":"SELL","type":"MARKET","quantity":"0.123"}`
`https://fapi.binance.com/fapi/fapi/v1/order{"symbol":"BTCUSDT","side":"BUY","type":"MARKET","quantity":"1"}`
**OKEX**
`https://www.okex.com/api/spot/v3/orders{"type": "limit", "side": "buy", "instrument_id": "BTC-USDT", "size":"0.001", "client_oid": "oktspot79", "price": "4638.51", "notional": "", "order_type": "3"}`
`https://www.okex.com/api/futures/v3/order{"client_oid": "y12233456","order_type":"1","instrument_id":"BTC-USD-180213","type":"1","price":"432.11","size":"2","match_price":"0"}`
**Karken**
`https://api.kraken.com/0/private/AddOrder{"pair":"XBTCHF", "type":"buy","ordertype":"market","volume":"0.123"}`
`https://futures.kraken.com/derivatives/api/v3/sendorder{"orderType":"lmt","symbol":"fi_xbtusd_180615","side":"buy","size":"1","limitPrice":"2321.2"}`
**Derbit**
`https://www.deribit.com/api/v2/public/get_order_book?instrument_name=BTC-22MAR20-5375-C`
`https://www.deribit.com/api/v2/private/buy{"instrument_name":"BTC-22MAR20-5375-P", "amount":"buy","type":"market","amount":"1"}`

**Comment by walec51** [comment]
whats the difference between a position and an order?
why would we have both for futures?

**Comment by douggie** [comment]
Conceptually it is like an amount, basically a running total of your orders.

Buy 10
Sell 5
Buy 4

My position would be 9. When trading spot you are buying and selling
currencies, so your position in your balance. For derivatives you are
buying and selling contracts. So you have two things, your position
(basically the running total of contracts your have bought and sold) and
your balance, the actual amount of currencies in your account.

**Comment by walec51** [comment]
if it should represent a collection of orders then I understand that this makes no sense:

```
Position extends Order
```

?

**Comment by douggie** [comment]
Correct, position should not extend order, the only real common items are the instrument and quantity. Position is the aggregation/running total of the fills.

**Comment by makarid** [comment]
You have a position only when you use leverage. That way your balance doesn't change(you don't sell bitcoin and get euro for example.). A position is opened with exchange's money, using your balance as collateral. An order is when you exchange X currency with Y currency which will change you balance.

**Comment by walec51** [comment]
@douggie if your thesis that orders for derivatives can be represented in the same way as orders for normal currency pairs then I'm in favor of your solution

however they way you prepared your pull request is not the way to go - reviewing 400 files is impossible and no one would upgrade to the new version because it would break every existing code that uses XChange

I think we should slice out one aspect at a time out of it and try to make the API backwards compatible

lets only start with the order `Order` and just do:

```
public abstract class Order implements Serializable {

  // remove:
  private final CurrencyPair currencyPair;
  // with
  private final Instrument instrument;

  // in all constructors in Order replace 'CurrencyPair currencyPair' with 'Instrument instrument`

  // provide these methods

  @Deprecated
  public CurrencyPair getCurrencyPair() {
    if (!(instrument instanceof CurrencyPair)) {
      throw new IllegalStateException("The instrument of this order is not a currency pair: " + instrument);
    }
    return (CurrencyPair) instrument;
  }

  public Instrument getInstrument() {
    return instrument;
  }

  // in the builder:

    public Builder currencyPair(CurrencyPair currencyPair) {
      this.instrument= currencyPair;
      return this;
    }

    
    public Builder instrument(Instrument instrument) {
      this.instrument= instrument;
      return this;
    }
}
```

if you apply my suggestion above correctly then you will not have to modify all the exchange implementations in this project

and you will be able to add support for placing future orders in `TradeService` `place*Order` in the few exchanges that support it

lets do this for a start, this should be a PR of no more then a dozen files

**Comment by walec51** [comment]
> You have a position only when you use leverage

is a position something specific to derivatives?
or can you have a position on a normal currency pair like BTC/USD?

**Comment by douggie** [comment]
Sounds good. Will make it backwardly compatible as I can there we still a few bits that needed changing like the serializer in the rock. Exchange

**Comment by walec51** [comment]
please make a separate PR and just focus on applying the Instrument class to Order in the way I described, try to keep the changes outside of xchange-core to a minimum

**Comment by douggie** [comment]
Hi Adam,

Think it will be a bit more than just the order class as I feel at a
minimum we should be able to do the following.

* Collect trade and depth data (changes to market data services)
* Place and cancel orders

But rather than refactoring the exchange implentations I will leave them
unchanged and just update xchange-core.

**Comment by mdvx** [comment]
One other thing on positions:
Derivatives contracts tend to have and expiry or end date (not perpetuals), which reset the position when it is hit.

Buy +10 June  -> positions June(10)
Sell -5 June   -> positions June(5)
Sell -3 September   -> positions June(5), September(-3)
# June 20th (3rd friday+1 day)  -> positions June(0), September(-3)
Buy +4 September    -> positions September(1)

**Comment by walec51** [comment]
@douggie 

please focus only on: Place orders

in the first step - it is the most crucial one

we will hit a wall again if you bring +100 file pull request

this:
* Collect trade and depth data (changes to market data services)
* cancel orders

we can do in other subsequent PRs after we have ironed out the place orders PR

**Comment by douggie** [comment]
No problem. Will do them one by one.

**Comment by Spyros-Stylianou** [comment]
Just come across this while trying to integrate a futures exchange. What work remains on this issue? Looks like we've now introduced the Instrument class which can be extended to support Options/Futures.

Can we change the Services in core to take Instrument instead of CurrencyPair or does some further work need to be done before this happens. It looks like we will have to go back and ensure existing implementations aren't referencing the base/counter fields of CurrencyPair. Or is there something else which should be done as the next step.

**Comment by douggie** [comment]
Hi Spyros,

I did not have time to finish it, if you take a look at the PR
https://github.com/knowm/XChange/pull/3473 it has an example in it for
okex, if you implement inline with the comments, you should be good to go.
Or feel free to update the PR.

**Comment by Viserius** [comment]
Any news on futures support?

**Comment by lythanh24** [comment]
any update on futures support?
